packages/apps/plugins/plugin-navtree/src/NavTreePlugin.tsx

```diff
@@ -39,13 +43,70 @@ export type NavTreePluginProvides = SurfaceProvides &
   TranslationsProvides;
 
 export const NavTreePlugin = (): PluginDefinition<NavTreePluginProvides> => {
+  const longestPaths = new Map<string, string[]>();
+  const state = deepSignal<{ root?: TreeNode }>({});
   let graphPlugin: Plugin<GraphProvides> | undefined;
 
+  const filterLongestPath: NodeFilter = (node, connectedNode): node is Node => {
+    const longestPath = longestPaths.get(node.id);
+    if (!longestPath) {
+      return false;
+    }
+
+    if (longestPath[longestPath.length - 2] !== connectedNode.id) {
+      return false;
+    }
+
+    return true;
+  };
+
   return {
     meta,
     ready: async (plugins) => {
       graphPlugin = resolvePlugin(plugins, parseGraphPlugin);
 
+      effect(() => {
+        longestPaths.clear();
+        graphPlugin?.provides.graph.traverse({
+          visitor: (node, path) => {
+            const longestPath = longestPaths.get(node.id)?.length ?? 0;
+            if (longestPath < path.length) {
+              longestPaths.set(node.id, path);
+            }
+
+            // TODO(wittjosiah): Factor out.
+            let shortcut: string | undefined;
+            if (typeof node.properties.keyBinding === 'object') {
+              const availablePlatforms = Object.keys(node.properties.keyBinding);
+              const platform = getHostPlatform();
+              shortcut = availablePlatforms.includes(platform)
+                ? node.properties.keyBinding[platform]
+                : platform === 'ios'
+                  ? node.properties.keyBinding.macos // Fallback to macos if ios-specific bindings not provided.
+                  : platform === 'linux' || platform === 'unknown'
+                    ? node.properties.keyBinding.windows // Fallback to windows if platform-specific bindings not provided.
+                    : undefined;
```

Comment:
Could this logic possibly live somewhere else? Feels slightly uncomfortable in NavTreePlugin, maybe at least worth a to-do.

Reply:
yeah, there's a factor out todo above this. I'm expecting this to move over to a search/commands plugin in the near future.